packages/ra-data-simple-rest/src/index.ts

```diff
@@ -25,72 +96,99 @@ import {
  * CREATE       => POST http://my.api.url/posts
  * DELETE       => DELETE http://my.api.url/posts/123
  */
-export default (apiUrl, httpClient = fetchUtils.fetchJson) => {
+export default (
+    apiUrl: string,
+    httpClient: HTTPClient = fetchUtils.fetchJson
+) => {
     /**
      * @param {String} type One of the constants appearing at the top if this file, e.g. 'UPDATE'
      * @param {String} resource Name of the resource to fetch, e.g. 'posts'
      * @param {Object} params The data request params, depending on the type
      * @returns {Object} { url, options } The HTTP request parameters
      */
-    const convertDataRequestToHTTP = (type, resource, params) => {
+    const convertDataRequestToHTTP = (
+        type: string,
+        resource: string,
+        params: RequestParams
+    ) => {
         let url = '';
-        const options = {};
+        const options: RequestOptions = {};
         switch (type) {
             case GET_LIST: {
-                const { page, perPage } = params.pagination;
-                const { field, order } = params.sort;
+                const params_ = params as GetListRequestParams;
```

Comment:
This shouldn't be necessary as TypeScript knows how to infer type on discimiinated unions (cf http://www.typescriptlang.org/docs/handbook/advanced-types.html#discriminated-unions)

Reply:
Thank you for reviewing.
I think "params" need to have a property to identify the type to do that (like s.kind in cf).

Reply:
that's the `type` on which you `switch`

Reply:
`type` is not a property of `params`, so I think "discimiinated unions" pattern can't be applied this.

Reply:
yes, it should work - read again the doc I pointed to. We've already used it in the react-admin code base, in the reducers.

Reply:
It means changing to like following code?

```ts
    /**
     * @param {String} req.type One of the constants appearing at the top if this file, e.g. 'UPDATE'
     * @param {String} req.resource Name of the resource to fetch, e.g. 'posts'
     * @param {Object} req.params The data request params, depending on the type
     * @returns {Object} { url, options } The HTTP request parameters
     */
    const convertDataRequestToHTTP = (req: RequestData) => {
        let url = '';
        const options: RequestOptions = {};
        switch (req.type) {
            case GET_LIST: {
                const { page, perPage } = req.params.pagination;
                const { field, order } = req.params.sort;
```